cluster-autoscaler/proposals/cp_agnostic_ca.md

```diff
@@ -0,0 +1,158 @@
+	mutex.Lock()
+	defer mutex.Unlock()
+	if cloudProviderFactory != nil {
+		klog.Fatalf("Cloud provider %q was registered twice")
+	}
+	cloudProviderFactory = cpf
+}
+
+func BuildExternalCloudProvider(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+	mutex.Lock()
+	defer mutex.Unlock()
+	if cloudProviderFactory != nil {
+		klog.Fatalf("external cloudProvider not registered")
+	}
+	return cloudProviderFactory(opts, do, rl)
+}
+```
+
+- Modify buidler_all.go
+```go
+// +build  !external !gce,!aws,!azure,!kubemark,!alicloud,!magnum,!digitalocean,!clusterapi
+package builder
+
+import (
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/alicloud"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/azure"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/baiducloud"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/clusterapi"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/digitalocean"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/external"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/gce"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/magnum"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/packet"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+)
+
+// AvailableCloudProviders supported by the cloud provider builder.
+var AvailableCloudProviders = []string{
+	cloudprovider.AwsProviderName,
+	cloudprovider.AzureProviderName,
+	cloudprovider.GceProviderName,
+	cloudprovider.AlicloudProviderName,
+	cloudprovider.BaiducloudProviderName,
+	cloudprovider.MagnumProviderName,
+	cloudprovider.DigitalOceanProviderName,
+	cloudprovider.ExternalCloudProviderName,
+	clusterapi.ProviderName,
+}
+
+// DefaultCloudProvider is GCE.
+const DefaultCloudProvider = cloudprovider.GceProviderName
+
+func buildCloudProvider(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+	switch opts.CloudProviderName {
+	case cloudprovider.GceProviderName:
+		return gce.BuildGCE(opts, do, rl)
+	case cloudprovider.AwsProviderName:
+		return aws.BuildAWS(opts, do, rl)
+	case cloudprovider.AzureProviderName:
+		return azure.BuildAzure(opts, do, rl)
+	case cloudprovider.AlicloudProviderName:
+		return alicloud.BuildAlicloud(opts, do, rl)
+	case cloudprovider.BaiducloudProviderName:
+		return baiducloud.BuildBaiducloud(opts, do, rl)
+	case cloudprovider.DigitalOceanProviderName:
+		return digitalocean.BuildDigitalOcean(opts, do, rl)
+	case cloudprovider.MagnumProviderName:
+		return magnum.BuildMagnum(opts, do, rl)
+	case packet.ProviderName:
+		return packet.BuildPacket(opts, do, rl)
+	case clusterapi.ProviderName:
+		return clusterapi.BuildClusterAPI(opts, do, rl)
+	case cloudprovider.ExternalCloudProviderName:
```

Comment:
can you elaborate how you plan to test this new provider ?

Reply:
Testing new provider (private provider) is outside the purview of CA.

Reply:
Sorry, but I disagree any piece of code in the CA should be covered with tests. You could have a fake external provider.

Reply:
Agreed, There should be testcases which cover the wiring of external_provider.
Whether or not to have a test to check execution of dummy_cp can be discussed.